### ISSUE TITLE
getnetworkhashps seems to be unrelated to actual network hash.

### ISSUE THREAD

**Issue by seanPhill** [Veil-Project/veil#340]
When checking getnetworkhashps the result seems to be nowhere near a believable value.

testnet
```
getnetworkhashps
17.08739622641509

getnetworkhashps
￼16.4808560197991
```
mainnet
```
getnetworkhashps
16.51547897196262

getnetworkhashps
16.7731179608773
```
That's crazy low, isn't it?
But the difficulty is high.

testnet
```
 "difficulty": 407361.8355413844,
```
mainnet
```
"difficulty": 1739806.144906988,
```

**Comment by CaveSpectre11** [comment]
This actually isn't a bug.  It's a little misleading due to the fact that the difficulty checks happen to be done when the block was a PoS block rather than a PoW block.  The `getnetworkhashps` is an average per second over the last 120 blocks (without using parameters to change the time) and is based solely on the timestamp of the two blocks and the chainwork.  PoS uses a different mechanism for it's "work" target, so the chainwork is not going to accurately reflect the work.

What you end up with is a hash per second that doesn't really present any useful information, since it's a fruit salad of PoW apples vs. PoS oranges.  But mathematically, it's correct.

e.g.
PoS blocks 306022 to 306026 had difficulties over 44.6 mil; but the chain work went from 0x1964727a to 0x19648f67; a difference of 7405 over a period of 281 seconds.  7405/281 = 26.35

A more accurate reflection of the hashPS is calculated by looking at a stretch of PoW blocks.  e.g. 306015 to 306018.   difficulty at ~4k, and a hashPS of 11.71.

Comparatively; 169998 through 170000; which has a higher PoW difficulty, comes up with a hashPS of 13.54

So while it looks strange to see such low Hash PS with such high difficulties; what you're really seeing is a very disconnected average of PoS vs. PoW.

**Comment by Rock-N-Troll** [comment]
So essentially the change you're suggesting is probably the best course of action for getting useful data out of this api call. I agree with that.

**Comment by CaveSpectre11** [comment]
So you want the api call for getting the network hash per second to give you something more useful then the network hash per second?

**Comment by Rock-N-Troll** [comment]
As you've mentioned, removing PoS values from the calculation would be a more accurate reflection of network hashPS

**Comment by CaveSpectre11** [comment]
Fair enough; it would be exponentially more complex; but definitely possible  (instead of taking the numbers from two blocks and averaging; it would need to walk the chain between the two blocks and total the hash rates for PoW blocks, and discounting the PoS blocks (as well as the time rates too).

But that doesn't solve the premise of the original issue:  That the high [PoS] difficulty didn't make sense to the low hash rate.  The hash rate would probably adjust a little bit when removing the PoS blocks from it; but it wouldn't make too much difference at the low historical difficulties. (11 vs 15 hashes per second type thing).

I would suggest a better approach is to leave networkhashps alone, and focus on the "difficulty" report; having it reflect if the difficulty is the PoS or PoW number, like the explorer does:

![image](https://user-images.githubusercontent.com/36988814/62721363-09352e00-b9da-11e9-83f4-3ec39bf0052a.png)

**Comment by Rock-N-Troll** [comment]
I'm not sure that would necessarily be more complex. A queue with the X most recent PoW blocks wouldn't be exponentially more difficult to implement, unless I'm really missing something here. Also, I believe the hash PS is in Gigahash/per second or some factor that isn't immediately obvious.

**Comment by CaveSpectre11** [comment]
The logic currently is essentially:
` (block(tip)->chainwork minus block(y)->chainwork) / *block(tip)->timestamp minus block(y)->timestamp)`

There's no walking the chain.  To walk the chain, you're walking through and totaling up the chainwork deltas and timestamp deltas.  Which, from a logical standpoint is expontential... 4(0) to 4 times tip.height-y.height (0) if we're talking big 0 notation.  

And like I said before; it won't make much difference in the number; and won't solve the source of the issue... that the number looked greatly different then the difficulty.  If you looked at the block I took the screen grab of.. you'll see a difficulty of 31,597,403 and a hash PS that's reflective of a PoW difficulty 1/7000 of the PoS difficulty.

Basically:  "When checking getnetworkhashps the result seems to be nowhere near a believable value."; which doesn't change because the hash PS is fairly accurate(*).  (accurate in a +/- 100%; compared to an expectation of a value that would be 7000 times the accurate value).

**Comment by Rock-N-Troll** [comment]
(block(tip)->chainwork minus block(y)->chainwork) / *block(tip)->timestamp minus block(y)->timestamp)

This formula is wrong.

https://en.bitcoin.it/wiki/Difficulty#What_network_hash_rate_results_in_a_given_difficulty.3F

https://bitcoin.stackexchange.com/questions/1770/how-can-i-measure-the-total-hashing-power-of-the-bitcoin-network/1774#1774

Network hashrate calculated using formula: H ~= h / t , where t is time that took to find X number of blocks and h is approximate number of hashes it should have taken to solve X number of blocks, h = X * (D * 2**256 / (0xffff * 2**208))"

It's not supposed to be chainwork, it's supposed to be number of blocks * (difficulty * 2^256/difficulty offset)

So unless these values somehow reduce to chainwork, I think the formula in use is wrong.

**Comment by CaveSpectre11** [comment]
That could be possible; honestly I don't remember the exact formula, and what I was explaining was greatly simplified... just illustrating it's looking at two blocks only, and not deciphering the chain to rule out PoS blocks.

But the premise of the issue still needs to be solved.  Whatever the formula is; a hashes per second number when the PoW difficulty is 6000 will never make sense when being compared to a PoS difficulty of 1.7m.  That initial misunderstanding needs a solution.

**Comment by Rock-N-Troll** [comment]
Right, which is why PoS must be disregarded from this calculation. Only look at PoW difficulty and PoW variables.

**Comment by CaveSpectre11** [comment]
Correct; but we're still bouncing around on the same thing.  We really need Sean's input.

> That's crazy low, isn't it?
> But the difficulty is high.

The difficulty is high because it was the PoS difficulty that was being looked at.  _That_ is the root of the problem first and foremost.  You need to solve the issue of comparing a PoS difficulty to the hashes per second...

Yes, perhaps the hashes per second is not accurate; but that's kinda a lesser thing to this issue as written.  If it did reflect the PoS difficulty as a PoW difficulty; than the premise of it being "too low for the difficulty" wouldn't have been noticed.... because it wouldn't have been.   For this, the "crazy low", is because the wrong difficulty was being looked at by the user.

**Comment by ianGre** [comment]
I would agree with `getnetworkhashps` not to get mixed up with PoS, and to do the PoW hash rate only, but maybe there will still be problems because of the different PoW algorithms.
